feat: create header component

gllherme opened the pull request:
**Sumary**
Creates the header component with flexibility for future changes and/or additions. The additions include the main header, its two controlled states (Guest and Logged In), and its subcomponents with mocked values.

- Issue: #3

**Changes**
- `src/components/Header/index.tsx`: Creates the main Header component. The Logged In/Guest state is controlled by the "isLogged" constant which determines wheter `<LoggedInComponents/>` is rendered or not.

- `src/components/Header/HeaderSearch/index.tsx`: Creates the HeaderSearch component. This is a controlled input that is controlled in the parent component by the `handleChange` function and activated by pressing the `Enter` key.

- `src/components/Header/HeaderNavItem/index.tsx`: Creates the NavItem component that has three styles: default, hover and active. The "isActive" constant checks if the NavItem href is the current page, if true, it makes the NavItem active.

- `src/components/Header/AccountNotification/index.tsx`: Creates the AccountNotification component. This component simply renders an icon that can bet set through props and the number of notifications.

- `src/components/Header/AccountMenu/index.tsx`: Creates the AccountMenu component. This component shows the username and avatar of the logged in user.

- `src/components/Header/HeaderBanner/index.tsx`: Creates three banner components for warnings, announcements and twitch lives notifications. Each banner have a "message", href and "onClick" prop for different uses.

**Images**

Guest header:

![image](https://github.com/YARC-Official/yarg.in/assets/53663956/25f4ea76-2688-4490-a234-61c31269e9bb)

Logged In header: 

![image](https://github.com/YARC-Official/yarg.in/assets/53663956/57835e9e-052d-4e34-86bd-5dca196e37d5)

Header with banners: 

![image](https://github.com/YARC-Official/yarg.in/assets/53663956/9f9d0625-54c3-4dfd-95a5-b6368d37cda1)

Behavior:

![2024-02-04 03-16-14](https://github.com/YARC-Official/yarg.in/assets/53663956/d5ddedaa-0503-4ffb-8162-9bfa9e5b729a)

gllherme committed:
feat: create header component

creates the main header component and it's subcomponents: NavItem, HeaderSearch,
AccountNotification, AccountMenu and HeaderBanner.

gllherme committed:
refactor: crypyo UUID instead of index on map

kaduwaengertner commented:
Really awesome, one small thing we forgot to say is that the header going to be fixed on top, so it matches with the other component style

Other than that looks pretty amazing

Luisgustavom1 commented on src/components/Header/index.tsx:
```diff
@@ -1,54 +1,103 @@
+    <ul>  
+      <NavItem label="Home" href="/" />
+      {navigation.map(({ href, name, redirect }) => (
+        <NavItem key={crypto.randomUUID()} href={href} label={name} target={redirect ? "_blank" : "_self"} /> 
```
The key should be match between renders, so generating random number with crypto on the fly it's not good practice, because react will lose the reference of the list item and bugs may appear.

My suggestion is define a unique id to each element of `navigation`, as it is not yet dynamic, so you have full control over whether you define an id or a unique attribute.

If you want to read more about https://react.dev/learn/rendering-lists#:~:text=Similarly%2C%20do%20not,on%20the%20data.

edmolima commented on src/components/Header/AccountMenu/index.tsx:
```diff
@@ -0,0 +1,20 @@
+import Image from "next/image";
+import styles from "./account-menu.module.css";
+
+const AccountMenu = () => {
```
You could use destructuring and create a default value when this data doesn't have filled

Example:

```
export default (props) => {
    const { username = "Kadoodle" } = props
    return (
        <h1>Hello {username}!</h1>;
    )
}
```

This will ensure that this component could be extensible

guim0 replied:
@edmolima 
just a question, if the solution was like this:

```
export default ({username = "Kadoodle" } : IAccountMenu) => {
    return (
        <h1>Hello {username}!</h1>;
    )
}
```

Could work?

edmolima replied:
Hey @guim0, yes this is absolutely work

I told to you use like receive "props" because I use this notation:

```
export const Example: FC<IExampleProps> = (props) => {
}
```

But anyway, you can use like this

```
export default ({username = "Kadoodle" } : IAccountMenu) => {
    return (
        <h1>Hello {username}!</h1>;
    )
}
```

I mean, do not use just export default anonymous function, is prefere to use export default NameComponent
this should be work better

edmolima commented on src/components/Header/AccountMenu/account-menu.module.css:
```diff
@@ -0,0 +1,42 @@
+
+  background: none;
+
+  border: 1px solid #0F121D;
```
There is also a possibility of leaving all these hexadecimal values in variables

if it is not necessary to separate it into a file, it makes sense to leave it at the top of the file

matheusmorett2 commented on src/components/Header/HeaderSearch/header-search.module.css:
```diff
@@ -0,0 +1,52 @@
+   background-size: var(--_icon-size);
+}
+
+.search:focus {
+   color: #676B7D;
+   background-color: #FFFFFF;
+}
+
+.search::-webkit-input-placeholder { /* WebKit, Blink, Edge */
+   color:#A6ACCA;
+}
+.search:-moz-placeholder { /* Mozilla Firefox 4 to 18 */
+   color:#A6ACCA;
+   opacity: 1;
+}
+.search::-moz-placeholder { /* Mozilla Firefox 19+ */
+   color:#A6ACCA;
+   opacity: 1;
+}
+.search:-ms-input-placeholder { /* Internet Explorer 10-11 */
+   color:#A6ACCA;
+}
+.search::-ms-input-placeholder { /* Microsoft Edge */
+   color:#A6ACCA;
+}
+.search::placeholder { /* Most modern browsers support this now. */
+   color:#A6ACCA;
+}
```
I also think we should use variables instead of manually have the hexadecimal for each color definition

matheusmorett2 commented on src/components/Header/HeaderNavItem/index.tsx:
```diff
@@ -0,0 +1,22 @@
+  const isActive = pathname === href;
+
+  return (
+    <a href={href} target={target}>
+      <li className={isActive ? styles.active : styles.default}>{label}</li>
+    </a>
```
We can have `<li><a></a></li>` if we wanna follow the best HTML semantics practice. See https://developer.mozilla.org/en-US/docs/Learn/HTML/Introduction_to_HTML/Document_and_website_structure.

gllherme replied:
you are absolutely right, going to fix it.

matheusmorett2 commented on src/components/Header/HeaderBanner/index.tsx:
```diff
@@ -0,0 +1,42 @@
+
+const WarningBanner: React.FC<Props> = ({ href, message, onClick }) => {
+  return (
+    <div className={[styles.banner, styles.warning].join(" ")} onClick={onClick}>
+     <AlertTriangle className={styles.icon} /><a href={href} target="_blank">{message} Click here to know more.</a>
+    </div>
```
We should avoid using `onClick` on `<div>`, instead a `<button>` would be a better option for this case. See https://www.w3schools.com/accessibility/accessibility_buttons_links.php and https://www.yanandcoffee.com/2020/08/10/button-vs-div/ if you wanna understand how to still use a `<div>` but with a button role.

gllherme replied:
great point, i'm going to add `role="button"` to this div since it's not really a button. thanks!

matheusmorett2 commented on src/components/Header/AccountMenu/index.tsx:
```diff
@@ -0,0 +1,20 @@
+  return ( 
+    <section className={styles.container} >
+      <Menu className={styles.icon} />
+      <span className={styles.greeting}>
+        <span className={styles.username}>{username}</span>
+      </span>
```
why having a <span> inside a <span>? could this be just one with the 2 classes?

guim0 replied:
I think the purpose is to give a difference to the username text but add the "Hey, " 
![image](https://github.com/YARC-Official/yarg.in/assets/59894160/75feed67-0cb0-4058-96b6-c5581a629fe4)
this could solve it:

on the component:
 ```
 <p className={styles.greeting}>
        Hey, <span>{username}</span>
</p>
```

on CSS:

```
.greeting {
  font: 16px;
  font-weight: 400;
  color: #717F9E;
  
> span {
 font-weight: 600;
  color: #FFFFFF;
}


```

matheusmorett2 replied:
I see!! I missed that content on CSS.

gllherme replied:
> I think the purpose is to give a difference to the username text but add the "Hey, " ![image](https://private-user-images.githubusercontent.com/59894160/302131809-75feed67-0cb0-4058-96b6-c5581a629fe4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwODAxNTksIm5iZiI6MTcwNzA3OTg1OSwicGF0aCI6Ii81OTg5NDE2MC8zMDIxMzE4MDktNzVmZWVkNjctMGNiMC00MDU4LTk2YjYtYzU1ODFhNjI5ZmU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDIwNTA1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkMjQwNjY2YzU5YjA0ODM4MzlkMDljNmU4M2I5OWUyOGUyMzAzMTliOTRjNDJhNjFlMTdhYjRlYjQ5NDc4NTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AZsF9iImsvswHnMiBkjKfdh_O2_fYasDX2tPrNmcxu0) this could solve it:
> 
> on the component:
> 
> ```
> <p className={styles.greeting}>
>        Hey, <span>{username}</span>
> </p>
> ```
> 
> on CSS:
> 
> ```
> .greeting {
>   font: 16px;
>   font-weight: 400;
>   color: #717F9E;
>   
> > span {
>  font-weight: 600;
>   color: #FFFFFF;
> }
> ```

that's it, the greeting and username have different colors and font-weight, thus explaining the nested spans but your solution has better html semantic. going to improve it. thanks!

gllherme committed:
fix: makes the header fixed

gllherme committed:
refactor: improves html semantic and accessibility

gllherme committed:
refactor: creates AccountMenu props and default values

gllherme committed:
refactor: HeaderSearch confirmation on click and better styling

guim0 reviewed (approved):
LGTM

mastercoks commented on src/assets/Icons/MagnifyingGlass.svg:
recomendo utilizar a lib https://lucide.dev/icons/search

gllherme replied:
sim verdade, salvei esse svg antes de ver a lib e esqueci de apagar depois, vou corrigir.

mastercoks commented on src/components/Header/AccountMenu/index.tsx:
```diff
@@ -0,0 +1,26 @@
+
+const usernamePlaceholder = "Kadoodle"
+
+const AccountMenu: React.FC<Props> = ({ username = usernamePlaceholder, avatar = AvatarPlaceholder }) => {
```
eu removeria esses campos default e passaria eles na chamada desse componente

gllherme replied:
coloquei esses valores default pra mockar o username e o avatar e depois remover quando for feita a integração com o serviço, mas realmente não faz muito sentido adicionar algo pra remover depois. vou fazer essa correção.

mastercoks commented on src/components/Header/HeaderBanner/index.tsx:
minha sugestão seria juntar tudo em um único componente `HeaderBanner` e passar uma prop `type: 'warning' | 'announcement' | 'live'`

``` typescript
import { AlertTriangle, LucideIcon, Megaphone } from "lucide-react";
import styles from "./style.module.css";
import { Twitch } from "@/assets/Icons";

type BannerType = "warning" | "announcement" | "live";

type Props = {
  type:  BannerType;
  href?: string;
  message: string;
  onClick?: () => void;
};

const bannerTypes: Record<BannerType, {icon: LucideIcon, style: string}> = {
  announcement: {
    icon: Megaphone,
    style: styles.announcement
  },
  live: {
    icon: Twitch,
    style: styles.live
  },
  warning: {
    icon: AlertTriangle,
    style: styles.warning
  }
}

const HeaderBanner: React.FC<Props> = ({ type, href, message, onClick }) => {
  const {icon: Icon, style} = bannerTypes[type];

  return (
    <div className={[styles.banner, style].join(" ")} onClick={onClick} role="button">
     <Icon className={styles.icon} /><a href={href} target="_blank">{message} Click here to know more.</a>
    </div>
  );
};
```

algo nesse sentido

gllherme replied:
excelente sugestão, minha implementação tem muito código repetido, vou refazer esse componente.

gllherme committed:
refactor: refactors the entire HeaderBanner component

removes repeated code, facilitates adition of new banner types and fixes minor positioning issue

gllherme committed:
refactor: remove unused svg

gllherme committed:
refactor: removes unnecessary default values from AccountMenu
